### DISCUSSION TITLE
Can you create custom fields?

### DISCUSSION THREAD

**Discussion by TafkaMax** [codeaffen/phpipam-ansible-modules#79] [Q&A] [answered]
Can you create custom fields using these modules?

**Comment by cmeissner** [comment, marked as answer]
Hi @TafkaMax 

Thank you for bringing this question to our attention. Short answer is no - there is no module to manage custom fields.

I investigate this topic a bit further and come to the conclusion that there is no API endpoint to manage custom fields in phpipam yet.

You can ask for available API endpoints within `phpypam` library as followed:

```
$ python
Python 3.9.11 (main, Mar 21 2022, 19:15:19) 
[GCC 11.2.1 20220127 (Red Hat 11.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import phpypam
>>> pi = phpypam.api(url='https://localhost:8443', app_id='ansible', username='admin', password='ipamadmin', ssl_verify=False)
>>> pi.controllers()
{'sections', 'vrf', 'tools/racks', 'subnets', 'tools/nat', 'tools', 'vlan', 'addresses', 'tools/nameservers', 'tools/locations', 'tools/scanagents', 'folders'}
``` 

To request all API endpoints directly from a phpipam installation you can use `cURL` as followed:

```
$ TOKEN=$(curl -sk -XPOST -u admin:ipamadmin --basic https://localhost:8443/api/ansible/user | jq -r '.data.token')
$ curl -sk -XOPTIONS -H "token: ${TOKEN}" https://localhost:8443/api/ansible | jq '.data.controllers[].href'
"/api/ansible/sections/"
"/api/ansible/subnets/"
"/api/ansible/folders/"
"/api/ansible/addresses/"
"/api/ansible/vlan/"
"/api/ansible/vrf/"
"/api/ansible/tools/nameservers/"
"/api/ansible/tools/scanagents/"
"/api/ansible/tools/locations/"
"/api/ansible/tools/racks/"
"/api/ansible/tools/nat/"
"/api/ansible/tools/"
```

There is also no trace of a controller for custom fields. See [phpipam documentation](https://phpipam.net/api/api_documentation/) and [api source code files](https://github.com/phpipam/phpipam/tree/master/api/controllers).

What options do you have now. 

1. Open a feature request at [phpipam/phpipam](https://github.com/phpipam/phpipam/issues/new?assignees=&labels=feature+request&template=feature_request.md&title=)
2. Implement the API endpoint on your on and open a PR in phpipam/phpipam repo

If the API endpoint is available the development of a module in this repo can be started.

Unfortunately both issues and PR are abundantly available in phpipam project. So I would expect that neither option 1) nor 2) would bring the new API endpoint asap in the project.

Please excuse me for not being able to give you any other answer.

**Comment by TafkaMax** [comment]
Thank you for this information.
